Source/Diagnostics/WarpXOpenPMD.cpp

```diff
@@ -0,0 +1,473 @@
+  const amrex::Box& global_box = geom.Domain();
+  auto global_size = getReversedVec(global_box.size());
+  // - Grid spacing
+  std::vector<double> grid_spacing = getReversedVec(geom.CellSize());
+  // - Global offset
+  std::vector<double> global_offset = getReversedVec(geom.ProbLo());
+  // - AxisLabels
+#if AMREX_SPACEDIM==3
+  std::vector<std::string> axis_labels{"x", "y", "z"};
+#else
+  std::vector<std::string> axis_labels{"x", "z"};
+#endif
+
+  // Prepare the type of dataset that will be written
+  openPMD::Datatype datatype = openPMD::determineDatatype<amrex::Real>();
+  auto dataset = openPMD::Dataset(datatype, global_size);
+
+  auto series_iteration = m_Series->iterations[iteration];
+  series_iteration.setTime( time );
+
+  // Loop through the different components, i.e. different fields stored in mf
+  for (int icomp=0; icomp<ncomp; icomp++){
+
+    // Check if this field is a vector or a scalar, and extract the field name
+    const std::string& varname = varnames[icomp];
+    std::string field_name = varname;
+    std::string comp_name = openPMD::MeshRecordComponent::SCALAR;
+    bool is_vector = false;
+    for (const char* vector_field: {"E", "B", "j"}){
+        for (const char* comp: {"x", "y", "z"}){
+            if (varname[0] == *vector_field && varname[1] == *comp ){
+                is_vector = true;
+                field_name = varname[0] + varname.substr(2); // Strip component
+                comp_name = varname[1];
+            }
+        }
+    }
+
+    // Setup the mesh record accordingly
+    auto mesh = series_iteration.meshes[field_name];
+    mesh.setDataOrder(openPMD::Mesh::DataOrder::F); // MultiFab: Fortran order
+    mesh.setAxisLabels( axis_labels );
+    mesh.setGridSpacing( grid_spacing );
+    mesh.setGridGlobalOffset( global_offset );
+    setOpenPMDUnit( mesh, field_name );
+
+    // Create a new mesh record component, and store the associated metadata
+    auto mesh_comp = mesh[comp_name];
+    mesh_comp.resetDataset( dataset );
+    // Cell-centered data: position is at 0.5 of a cell size.
+    mesh_comp.setPosition(std::vector<double>{AMREX_D_DECL(0.5, 0.5, 0.5)});
+
+    // Loop through the multifab, and store each box as a chunk,
+    // in the openPMD file.
+    for (amrex::MFIter mfi(mf); mfi.isValid(); ++mfi) {
+
+      const amrex::FArrayBox& fab = mf[mfi];
+      const amrex::Box& local_box = fab.box();
+
+      // Determine the offset and size of this chunk
+      amrex::IntVect box_offset = local_box.smallEnd() - global_box.smallEnd();
+      auto chunk_offset = getReversedVec(box_offset);
+      auto chunk_size = getReversedVec(local_box.size());
+
+      // Write local data
+      const double* local_data = fab.dataPtr(icomp);
+      mesh_comp.storeChunk(openPMD::shareRaw(local_data),
+                           chunk_offset, chunk_size);
+    }
+  }
+  // Flush data to disk after looping over all components
+  //std::cout<<" this is optional "<<std::endl;
```

Comment:
```suggestion
```
ad this point it's not optional, since the pic simulation might now proceed and wants to use/manipulate the buffers that we registered in `storeChunk` again